src/main/java/tech/picnic/jolo/Loader.java

```diff
@@ -30,92 +30,77 @@
   }
 
-  /** Returns all objects loaded by this loader. */
-  public Stream<T> stream() {
-    return get().stream();
+  public static <T> LoaderBuilder<T> of(Entity<T, ?> mainEntity) {
+    return new LoaderBuilderImpl<>(mainEntity);
   }
 
-  /** Collects and then returns all objects loaded by this loader. */
-  public <R, A> R collect(Collector<? super T, A, R> collector) {
-    return stream().collect(collector);
+  /**
+   * Convenience function for improved readability when calling the loader as a parameter of {@link
+   * java.util.stream.Stream#collect(Collector)}. It returns the given loader.
+   */
+  public static <T> Loader<T> toLinkedObjectsWith(Loader<T> loader) {
+    return loader;
   }
 
-  /** Returns all objects loaded by this loader. */
-  public List<T> getList() {
-    return collect(Collectors.toList());
+  @Override
+  @SuppressWarnings("NoFunctionalReturnType")
+  public Supplier<ObjectGraph> supplier() {
+    return ObjectGraph::new;
   }
 
-  /**
-   * Returns all objects loaded by this loader. Note that the entity type must implement {@link
-   * Object#equals} and {@link Object#hashCode} appropriately in order for this method to return a
-   * correct result.
-   */
-  public Set<T> getSet() {
-    return collect(Collectors.toSet());
+  @Override
+  @SuppressWarnings("NoFunctionalReturnType")
+  public BiConsumer<ObjectGraph, Record> accumulator() {
+    return (objectGraph, record) -> {
+      for (Entity<?, ?> entity : entities) {
+        entity.getId(record).ifPresent(id -> objectGraph.add(entity, id, () -> entity.load(record)));
+      }
+      for (Relation<?, ?> relation : relations) {
+        objectGraph.add(relation, relation.getRelationLoader().apply(record));
+      }
+    };
   }
 
-  /**
-   * Returns the single object loaded by this loader, or throws an exception.
-   *
-   * @throws IllegalArgumentException if more than one object was loaded.
-   * @throws java.util.NoSuchElementException if the loader is empty.
-   */
-  public T getOne() {
-    return getOptional().orElseThrow(NoSuchElementException::new);
+  @Override
+  @SuppressWarnings("NoFunctionalReturnType")
+  public BinaryOperator<ObjectGraph> combiner() {
+    return (first, second) -> {
+      first.merge(second);
+      return first;
+    };
   }
 
-  /**
-   * Returns the single object loaded by this loader, if any.
-   *
-   * @throws IllegalArgumentException if more than one object was loaded.
-   */
-  public Optional<T> getOptional() {
-    Iterator<T> it = get().iterator();
-    if (it.hasNext()) {
-      T result = it.next();
-      if (it.hasNext()) {
-        throw new IllegalArgumentException("More than one entity was loaded");
+  @SuppressWarnings({"unchecked", "rawtypes", "NoFunctionalReturnType"})
+  @Override
+  public Function<ObjectGraph, List<T>> finisher() {
+    return objectGraph -> {
+      for (Relation relation : relations) {
+        relation.link(objectGraph.getObjectMapping(relation));
```

Comment:
If you make `objectGraph` responsible for doing the linking, that could simplify the code a bit. You would not have to instantiate this object mapping, and instead of using extra containers to store the intermediate results you could probably just have a `relation.link(left, right)` method that just accepts the left and right side of the relation. I think it would probably also shave a little bit of the performance overhead of this new implementation again.